Fix #415 - GROUP BY modifier WITH ROLLUP is treated as a syntax error and prevents export of SQL query results

niconoe- opened the pull request:
Fixes: #415

iifawzi commented on tests/data/parser/parseSelectGroupByErr.in:
```diff
@@ -0,0 +1 @@
+SELECT * FROM movie GROUP BY title WITH ROLLUP, year WITH ROLLUP;
```
Hi!

I think this's not a valid statement. `WITH ROLLUP` is most probably can be used once for the whole statement, not for each column, can you double-check that? I can't also see any references to such query in docs https://dev.mysql.com/doc/refman/8.0/en/group-by-modifiers.html.

If it's only used once, the `foreach` in the implementation is not needed then

niconoe- replied:
I know this is not a valid statement, that's why I added the suffix "Err" (for "Error") to the file as I wanted to test how it behaves when one's trying such invalid statement.

About the `foreach`, it is necessary IMO as it makes apply the "withRollup" modifier to every expression of the "GROUP BY".

niconoe- replied:
I see the "WITH ROLLUP" modifier as a kind of metadata on the "GROUP BY" section of the statement, so, as `GroupKeyword` instances are representing only expressions of such "GROUP BY", with 1 instance = 1 expression, I wanted to make sure that no matter the `GroupKeyword` instance we look at from the `SelectStatement` parsing, we can access to the same value for "withRollup" property (this property, as a metadata, is decoupled from each statement).

Using the `foreach` is probably performance killing, I could use a variable-by-reference instead so that switching from `false` to `true` if the parser encounters "WITH ROLLUP" token will affect all `GroupKeyword` instances we are returning with a single affectation rather than looping over all.

I also can try to provide a proposal with the status of the enabling of "WITH ROLLUP" on another level. My idea would be to say that `SelectStatement::$group` is no longer `array<GroupKeyword>` but rather `GroupByKeyword`: a new class that will hold the "withRollup" modifier as a boolean and the "expr" expressions that will be `array<GroupKeyword>`. But I'm a bit affraid of such changes as it has a heavy impact on `SelectStatement` that I can't measure yet...

iifawzi replied:
Hi, thank you for this detailed overview, and sorry for the late replies, I was just very busy in the past couple of days

> I see the "WITH ROLLUP" modifier as a kind of metadata on the "GROUP BY" section of the statement, so, as `GroupKeyword` instances are representing only expressions of such "GROUP BY", with 1 instance = 1 expression, I wanted to make sure that no matter the `GroupKeyword` instance we look at from the `SelectStatement` parsing, we can access to the same value for "withRollup" property (this property, as a metadata, is decoupled from each statement).
> 

do we really need to do this? I can't think of any benefits, in the end, `with rollup` will show more results in the select statements, and it has nothing to do with the columns, right? 

What I'd prefer now is supporting the keyword on the select statement level, maybe we can benefit from the logic we have for the end options 

https://github.com/phpmyadmin/sql-parser/blob/e1b21e1e87d626787580bf1f1aacab7e1d67a269/src/Statement.php#L372-L374

niconoe- replied:
Thank you for your answer. Using the `end_options` from the SELECT statement is a really good idea. If we use this, it will be more relevant in my opinion too. 
I'll try to take a look at this in the next day, as I think the parsing process will be obvious to change, but I have some doubts about the building process and I don't want to replace my bad algo in my proposal by another bad algo.

iifawzi replied:
Hi, thank you for the quick responses.

Take your time, and by the way, your proposal was never bad,  parser's logic is kinda complex and not well documented, everything can be done in thousands of ways. Hopefully in the future things can be more documented and simpler. 

Regarding the implementation, I think no build changes are needed, we already support the end options build logic, so most probably the build gonna work out-of-the-box

niconoe- replied:
> Regarding the implementation, I think no build changes are needed, we already support the end options build logic, so most probably the build gonna work out-of-the-box

The issue we have here is that "WITH ROLLUP" is definitly a group-by option, and not a select option. If we let the build as is, it will create such invalid statement:
```sql
SELECT title
FROM movies
GROUP BY title 
LIMIT 1 WITH ROLLUP;
```

Regarding of the build process is done over the SELECT (listing all clauses in orer and build them 1 by 1), it makes impossible to build a valid query as `end_options` will be managed long after the GroupKeyword clause.

niconoe- replied:
Ok, so, to fix this and make it buildable, I decided to add a property `with_rollup` in the SelectStatement class, which is false by default, but becomes true if the WITH ROLLUP modifier is declared.

I rolled back the changes on the GroupKeyword parser/builder and focused more on the Statement's one.

I added something that is a little bit hacky IMO as this required the "WITH ROLLUP" modifier to be considered as a keyword that deserves to be parsed (`$KEYWORD_PARSERS`), but as there are no class related (and we don't need, it's just a flag), I set the 'class' to `null`.

This allow the generic `build` method from the Statement class to still be used without any overwrite for the SelectStatement.

If that's too hacky and you find a better solution, please tell me as I'm starting to get rid of ideas :laughing:

iifawzi replied:
Missed that, you're right, we can't use end_options this way, and your solution is brilliant but In my opinion, we should keep doing things as we used to do with other options to avoid any hacky solutions, what I'd suggest is having a `group_options` that can be expected to be placed after the `group by `clause 

```PHP
    /**
     * @var array<string, int|array<int, int|string>>
     * @psalm-var array<string, (positive-int|array{positive-int, ('var'|'var='|'expr'|'expr=')})>
     */
    public static $GROUP_OPTIONS = [
        'WITH ROLLUP' => 1,
    ];
```

That can be then checked in the `Statement` this way: 

```PHP
 if (
      $this instanceof Statements\SelectStatement
      && $token->value === 'WITH ROLLUP'
  ) {
      // Handle group options in Select statement
      // See Statements\SelectStatement::$GROUP_OPTIONS
      $this->group_options = OptionsArray::parse($parser, $list, static::$GROUP_OPTIONS);
```

what do you think @niconoe-?

niconoe- replied:
I think that's awesome! Your proposal is keeping the same concept I tried to implement but without using the hack and actually using a class, so I think it could work. 

I'll give it a try after my lunch. Thanks a lot!

iifawzi commented:
Some changes are also needed here, so we can build the statements that include `WITH ROLLUP` correctly, most probably, you will need to append `WITH ROLLUP` to the end of the built statement. that's most probably as easy as checking if `WITH ROLLUP` is true, then add the token, otherwise don't add anything. 

https://github.com/phpmyadmin/sql-parser/blob/7315fc95248e88ef17033e8a392f0b332c1e11ae/src/Components/GroupKeyword.php#L122-L130

if you want to test this, you can add such a test in the `selectStatementTest.php` file

```PHP
    public function Testing(): void
    {
        $query = 'SELECT * FROM movie GROUP BY year WITH ROLLUP;';

        $parser = new Parser($query);
        $stmt = $parser->statements[0];

        $this->assertEquals($query, $stmt->build());
    }

```

niconoe- commented:
> Some changes are also needed here, so we can build the statements that include `WITH ROLLUP` correctly...

Thanks a lot! I was sure there was not only the "Parsing" process but also the "build" one, but I was unable to get it right 🙂 

I'll add this tomorrow and do some tests like you suggested. Thanks again!

williamdes commented on src/Components/GroupKeyword.php:
```diff
@@ -97,6 +101,14 @@ public static function parse(Parser $parser, TokensList $list, array $options =
                     $state = 0;
+                } elseif (($token->type === Token::TYPE_KEYWORD) && ($token->value === 'WITH ROLLUP')) {
+                    foreach ($ret as $previousExpr) {
+                        $previousExpr->withRollup = true;
```
I am not sure why but it feels weird to have this specific option as a property of this class
But regarding everything it's the only option I see

What do you think @iifawzi?
Do we often do that?

niconoe- replied:
I see what you mean. I thought about it last evening and even if "WITH ROLLUP" is a GROUP BY modifier, maybe the fact this modifier is on should be stored on the SelectStatement level. This because each `GroupKeyword` instance you have is representing not the `GROUP BY` but a single expression from this `GROUP BY`, and "WITH ROLLUP" modifier is not applicable per expression but per SELECT statement.

I could give it a try with this format if you prefer.

niconoe- replied:
Resolving this issue as I now worked on the resolution of WITH ROLLUP modifier in the SelectStatement, as advised.

niconoe- commented:
> Some changes are also needed here, so we can build the statements that include `WITH ROLLUP` correctly, most probably, you will need to append `WITH ROLLUP` to the end of the built statement. that's most probably as easy as checking if `WITH ROLLUP` is true, then add the token, otherwise don't add anything.

I updated the `build` method to manage the application of the "WITH ROLLUP" modifier whenever necessary, and I added 2 unit tests about it. Please tell me if you have other checks you want me to do.

williamdes commented on src/Components/GroupKeyword.php:
```diff
@@ -122,9 +135,22 @@ public static function parse(Parser $parser, TokensList $list, array $options =
+                return trim((string) $component->expr);
+            }, $component);
+            $expr = implode(', ', $expressions);
+        } else {
```
You can simply by removing the else keyword if the if always returns

williamdes reviewed (approved):
Thank you for killing the mutants!

niconoe- commented:
> Thank you for killing the mutants!

You're welcome. I think some of them can be killed by removing some sections of code. For instance, I spent some time on 
```
- if (($token->type === Token::TYPE_WHITESPACE) || ($token->type === Token::TYPE_COMMENT)) {
+ if (($token->type === Token::TYPE_WHITESPACE) && ($token->type === Token::TYPE_COMMENT)) {
```
but I realised that this was the same as 
```
if (false) {
```
so I wondered why it continued to pass the unit tests. And that's because if we don't ignore the whitespaces and the comments in the `GroupKeyword` class, they will be actually ignored also when parsing the expression with `Expression::parse(…)`. That means to me this `if` statement is useless on `GroupKeyword` class, except for small performance matters.

Anyway, I finally told myself that this was out of the context of the "WITH ROLLUP" modifier so I just ignored such mutants :smile:

williamdes commented:
> > Thank you for killing the mutants!
> 
> You're welcome. I think some of them can be killed by removing some sections of code. For instance, I spent some time on
> 
> ```
> - if (($token->type === Token::TYPE_WHITESPACE) || ($token->type === Token::TYPE_COMMENT)) {
> + if (($token->type === Token::TYPE_WHITESPACE) && ($token->type === Token::TYPE_COMMENT)) {
> ```
> 
> but I realised that this was the same as
> 
> ```
> if (false) {
> ```
> 
> so I wondered why it continued to pass the unit tests. And that's because if we don't ignore the whitespaces and the comments in the `GroupKeyword` class, they will be actually ignored also when parsing the expression with `Expression::parse(…)`. That means to me this `if` statement is useless on `GroupKeyword` class, except for small performance matters.
> 
> Anyway, I finally told myself that this was out of the context of the "WITH ROLLUP" modifier so I just ignored such mutants 😄

Yeah it's best to not mix PR because it makes them complicated to merge in terms of review.
But feel free to open a PR or an issue if you notice something that no test does cover

williamdes commented on src/Components/GroupKeyword.php:
```diff
@@ -122,9 +135,22 @@ public static function parse(Parser $parser, TokensList $list, array $options =
+            $expressions = array_map(static function ($component) use (&$withRollup) {
+                $withRollup = $component->withRollup;
+
+                return trim((string) $component->expr);
```
```suggestion
                return trim((string) $component->expr, ' ');
```
So you only trim spaces, is that not better ?

niconoe- replied:
See new tests from [new commit](https://github.com/phpmyadmin/sql-parser/pull/417/commits/19ffba31018419a9cefa74838d63456faa18cf40)

williamdes commented on src/Components/GroupKeyword.php:
```diff
@@ -122,9 +135,22 @@ public static function parse(Parser $parser, TokensList $list, array $options =
+            $expr = implode(', ', $expressions);
+        } else {
+            $withRollup = $component->withRollup;
+            $expr = trim((string) $component->expr);
```
```suggestion
            $expr = trim((string) $component->expr, ' ');
```

niconoe- replied:
I'll probably keep the `trim` as this. See new tests from [new commit](https://github.com/phpmyadmin/sql-parser/pull/417/commits/19ffba31018419a9cefa74838d63456faa18cf40).

niconoe- commented:
I don't understand why there is an issue with the codecov check: it tells me that AlterOperation.php is less covered than previously, but this file is not in my changes and AFAIK, there are no changes I made related to this file. Maybe I failed to rebase or something?

About the Mutation testing, all mutants that are applied on Statement.php file are not related to my changes, so I don't expect to kill them additionnaly in this PR, as I don't want to mix topics.

Therefore, I don't know how to proceed here. If that's required to pass all checks, could you give me some help?
Thanks a lot.

iifawzi commented on src/Parser.php:
```diff
@@ -190,6 +190,10 @@ class Parser extends Core
             'class' => 'PhpMyAdmin\\SqlParser\\Components\\GroupKeyword',
             'field' => 'group',
         ],
+        'WITH ROLLUP' => [
+            'class' => 'PhpMyAdmin\\SqlParser\\Components\\OptionsArray',
+            'field' => 'group_options',
+        ],
```
Would be better to name it `_GROUP_OPTIONS` instead of `WITH ROLLUP`

iifawzi commented on src/Statements/SelectStatement.php:
```diff
@@ -185,6 +190,10 @@ class SelectStatement extends Statement
             'GROUP BY',
             3,
         ],
+        'WITH ROLLUP' => [
```
here too

iifawzi commented on src/Statement.php:
```diff
@@ -62,6 +62,14 @@ abstract class Statement implements Stringable
      */
     public static $CLAUSES = [];
 
+    /**
+     * Options that can be given to GROUP BY component.
+     *
+     * @var array<string, int|array<int, int|string>>
+     * @psalm-var array<string, (positive-int|array{positive-int, ('var'|'var='|'expr'|'expr=')})>
+     */
+    public static $GROUP_OPTIONS = [];
+
```
why do we need this?

niconoe- replied:
Because of `$this->group_options = OptionsArray::parse($parser, $list, static::$GROUP_OPTIONS);`, which stands in the `Statement` class, so `static::$GROUP_OPTIONS` is not always defined if we don't provide this. It's detected by Psalm and PHPStan.

iifawzi reviewed (approved):
Brilliant hard work! well done, looks good to me
 
I think you will only need to squash the commits so @williamdes can do a final review and merge

niconoe- commented on src/Parser.php:
```diff
@@ -190,6 +190,10 @@ class Parser extends Core
+        '_GROUP_OPTIONS' => [
+            'class' => 'PhpMyAdmin\\SqlParser\\Components\\OptionsArray',
+            'field' => 'group_options',
+        ],
```
Done! Thanks :slightly_smiling_face:

niconoe- commented:
> Brilliant hard work! well done, looks good to me
> 
> I think you will only need to squash the commits so @williamdes can do a final review and merge

All fixed up! I also kept the first commits about `.out` files generations to compare before/after the implementation.

iifawzi commented:
> > I don't understand why there is an issue with the codecov check: it tells me that AlterOperation.php is less covered than previously, but this file is not in my changes and AFAIK, there are no changes I made related to this file. Maybe I failed to rebase or something?
> > About the Mutation testing, all mutants that are applied on Statement.php file are not related to my changes, so I don't expect to kill them additionnaly in this PR, as I don't want to mix topics.
> > Therefore, I don't know how to proceed here. If that's required to pass all checks, could you give me some help? Thanks a lot.
> 
> Erf, I still don't understand why the codecov is complaining. Maybe couple with the additional unit tests I provided on #421 it won't anymore, but TBH, I don't get the fact the coverage can differ on an unmodified file, especially a non-related one (that's not an abstract class which child stops to call a parent method, for instance…).
> 
> @iifawzi You got an idea about it, please?

sometimes codcov is weird but the diff looks good to me
cc @williamdes

niconoe- commented:
@williamdes / @iifawzi Do you need anything else to merge this into `master`? (Same question for #416 and #421 :smile: )

williamdes commented:
> @williamdes / @iifawzi Do you need anything else to merge this into `master`? (Same question for #416 and #421 😄 )

I will have a look soon, it's still unread in my sql-parser email folder 😄

williamdes reviewed (approved):
Thank you, and that's okay about codecov
